From 5cc9e62aad9fc922ab8f19c562c1e8713b7748c1 Mon Sep 17 00:00:00 2001
From: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Date: Tue, 28 May 2024 21:38:36 +0530
Subject: [PATCH] wifi: ath12k: set critical update for non-transmitting BSS

Add support in ath12k to handle critical update for both
transmitting and non-transmitting BSS.

Currently when arvif for non-Tx BSS has critical update, it
doesn't get set when the arvif is not up. Hence for link addition
critical update is not set. If the arvif is not up, current
set beacon will be bringing it up.

Set critical update for link addition even if arvif is not up.

Patch-depdendency: 855-01-ath12k-Set-critical-update-params-on-beacon-template.patch

Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/wmi.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index dae6743..44c5548 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -2922,7 +2922,11 @@ static void ath12k_wmi_bcn_fill_ml_info(struct ath12k_link_vif *arvif,
 			 */
 			if (link_conf->elemid_added || tx_link_conf->elemid_added)
 				set_bit(arvif_iter->vdev_id, &vdev_map_cat1);
-			if (link_conf->elemid_modified)
+			/* If arvif is not up, current set beacon will be bringing it up
+			 * So for link addition, set critical update even if arvif is
+			 * not up.
+			 */
+			if (link_conf->elemid_modified || !arvif_iter->is_up)
 				set_bit(arvif_iter->vdev_id, &vdev_map_cat2);
 		}
 	}
-- 
2.34.1

